heron/ckptmgr/src/java/com/twitter/heron/ckptmgr/backend/hdfs/HdfsBackend.java

```diff
@@ -0,0 +1,208 @@
+    }
+
+    return true;
+  }
+
+  @Override
+  public boolean restore(Checkpoint checkpoint) {
+    Path path = new Path(getCheckpointPath(checkpoint));
+
+    FSDataInputStream in = null;
+    CheckpointManager.SaveInstanceStateRequest state = null;
+    try {
+      in = fileSystem.open(path);
+      state =
+          CheckpointManager.SaveInstanceStateRequest.parseFrom(in);
+    } catch (IOException e) {
+      LOG.log(Level.SEVERE, "Failed to read", e);
+      return false;
+    } finally {
+      SysUtils.closeIgnoringExceptions(in);
+    }
+
+    checkpoint.setCheckpoint(state);
+
+    return true;
+  }
+
+  @Override
+  public boolean dispose(CheckpointManager.CleanStatefulCheckpointRequest request,
+                         String topologyName) {
+    String topologyCheckpointRoot = getTopologyCheckpointRoot(topologyName);
+    Path topologyRootPath = new Path(topologyCheckpointRoot);
+
+
+    if (request.hasCleanAllCheckpoints() && request.getCleanAllCheckpoints()) {
+      // Clean all checkpoint states
+      try {
+        fileSystem.delete(topologyRootPath, true);
+        if (fileSystem.exists(topologyRootPath)) {
+          return false;
+        }
+      } catch (IOException e) {
+        LOG.log(Level.SEVERE, "Failed to delete: " + topologyCheckpointRoot, e);
+        return false;
+      }
+    } else {
+      String oldestCheckpointPreserved = request.getOldestCheckpointPreserved();
+      try {
+        FileStatus[] statuses = fileSystem.listStatus(topologyRootPath);
+        for (FileStatus status : statuses) {
+          String name = status.getPath().getName();
+          if (name.compareTo(topologyCheckpointRoot) < 0) {
+            fileSystem.delete(status.getPath(), true);
+          }
+        }
+
+        // Do a double check. Now all checkpoints with smaller checkpoint id should be cleaned
+        statuses = fileSystem.listStatus(topologyRootPath);
+        for (FileStatus status : statuses) {
+          String name = status.getPath().getName();
+          if (name.compareTo(topologyCheckpointRoot) < 0) {
+            return false;
+          }
+        }
+
+      } catch (IOException e) {
+        LOG.log(Level.SEVERE, "Failed to clean to: " + oldestCheckpointPreserved, e);
+        return false;
+      }
+    }
+
+
+    return true;
+  }
+
+  /**
+   * Ensure the existence of a directory.
+   * Will create the directory if it does not exist.
+   *
+   * @param dir The path of dir to ensure existence
+   * @return true if the directory exists after this call.
+   */
+  protected boolean ensureDirExists(String dir) {
+    Path path = new Path(dir);
+
+    try {
+      fileSystem.mkdirs(path);
+      if (!fileSystem.exists(path)) {
+        return false;
+      }
+    } catch (IOException e) {
+      LOG.log(Level.SEVERE, "Failed to mkdirs: " + dir, e);
+      return false;
+    }
+
+    return true;
+  }
+
+  protected String getTopologyCheckpointRoot(String topologyName) {
+    return new StringBuilder()
```

Comment:
Prefer `String.format(..)` for static string template assembly. Here and elsewhere in this class and in LocalFS.